_posts/2019-08-13-snowplow-rdb-loader-r31-released-with-new-bad-rows/index.md

```diff
@@ -0,0 +1,141 @@
+
+While primary source of bad rows for Snowplow users is the Stream Enrich job that validates every event being processed by the pipeline, every microservice that makes up a Snowplow pipeline, including the RDB Loader, will emit a bad row if an attempt to process the event (in this case load it into Redshift) unsuccessful. We're happy to announce that Snowplow RDB Shredder is the first component of Snowplow pipeline released supporting the new format of bad rows.
+
+The RDB Shredder 0.15.0 can produce following types of bad rows:
+
+* `iglu:com.snowplowanalytics.snowplow.badrows/loader_parsing_error/jsonschema/1-0-0` - a generic loader error, caused if the event is not a valid "enriched event"
+* `iglu:com.snowplowanalytics.snowplow.badrows/loader_iglu_error/jsonschema/1-0-0` - an error raised by Iglu Client
+* `iglu:com.snowplowanalytics.snowplow.badrows/loader_runtime_error/jsonschema/1-0-0` - an unstructured error, raised by one of low-level shredding components, such as DynamoDB outage during cross-batch deduplication
+
+<h2 id="synthetic-duplicates-issue">2. Synthetic duplicates issue</h2>
+
+In [Snowplow R86 Petra][snowplow-r86] we introduced [an in-batch synthetic deduplication][synthetic-deduplication] step. This step identified rows of data that were distinct (i.e. have different event fingerprints), but that had the same `event_id`: this can occur if for example the Snowplow Javascript is executed in an environment where it cannot generate UUIDs e.g. by a robot scraping the website, for example. Where two events in a batch have the same event ID but different event fingerprints, they are assigned a new event ID to prevent collision, with the original event ID preserved in the `com.snowplowanalytics.snowplow/duplicate/jsonschema/1-0-0` context that is attached to any event that is mutated in this way.
+
+Recently, we discovered an anomaly in shredded data, where some events, which were clearly synthetic duplicates, were not "attached" to any context or unstructured event tables.
+Vice versa, some contexts and unstructured events had `root_id` values which did not correspond to any rows in `atomic.events` table, effectively resulting in "orphaned events".
+
+It turned out that this was caused by an issue with how Apache Spark caches parts of its execution DAG.
+After the shred job created an event object with new unique `event_id`, it caches this object and proceeds to the next step of creating an object containing all shredded entities with the same id.
+But if for some reasons there is not enough memory to store the initial cached object,  a new event object would be generated, along with new unique event id.
+This second event id would only be used only in the shredded entities as original event would already be written to HDFS, resulting in the `event_id` on the row in the events table not matching the `root_id` on the associated shredded tables.
```

Comment:
```suggestion
This second event id would only be used only in the shredded entities since the original event would already be written to HDFS, resulting in the `event_id` on the row in the events table not matching the `root_id` on the associated shredded tables.
```